docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+when its tag value matches `tag`; otherwise, it passes the reduced variant to
+`ElseOper`. This is precisely captured by the inference rule:
+
+```
+variant: Variant(RowCons("<TAG>", r, t))
+ThenOper: r => z
+ElseOper: Variant(t) => z
+------------------------------------------------ [variant_match]
+MatchTag(variant, "<TAG>", ThenOper, ElseOper):
+  (Variant(RowCons("<TAG>", r, t)),
+   Str,
+   r => z,
+   Variant(t) => z
+  ) => z
+```
+
+The above rule looks menacing. Here is an example of matching a record
+in the above example with the variable `log`:
+
+```tla
+IsDefined(eventAOrB) ==
+  LET ElseB(onlyB) ==
+    MatchTag(onlyB, "EventB", LAMBDA b: b.src /= "", LAMBDA elseValue: FALSE)
+  IN  
+  MatchTag(event, "EventA", LAMBDA a: a.val /= -1, ElseB)
+```
+
+The operator `ElseB` looks redundant, as we know that `onlyB` is a singleton
+variant. To this end, we introduce the operator `MatchOnly`:
+
+```tla
+MatchOnly(variant, ThenOper(_)) ==
+  \* fallback untyped implementation
+  ThenOper(variant)
+```
+
+For completeness, we give the inference rule for `MatchOnly`:
+
+```
+variant: Variant(RowCons("<TAG>", r, RowNil))
+ThenOper: r => z
+------------------------------------------------ [variant_match_only]
+MatchOnly(variant, ThenOper):
+  (Variant(RowCons("<TAG>", r, RowNil)),
+   r => z
+  ) => z
+```
+
+It looks like the solution with `MatchTag` and `MatchOnly` introduce a lot of
+boilerplate. However, this is probably the best solution that we can have,
+unless we can extend the grammar of TLA+.
+
+### 5.3. Changes in the model checker
+
+Having precise types for variants, we have two options:
+
+ 1. Keep the current encoding, that is, a variant is encoded as a super-record
+ that contains all possible fields of the member records. The type checker
+ will guarantee that we do not access the fields of the super-record that are
+ not present in the actual record type.
+
+ 1. Implement the
+ [suggestion by Shon Feder](https://github.com/informalsystems/apalache/discussions/789#discussioncomment-1592118).
+
+Although Option 2 looks nicer, we prefer keeping Option 1. The reason is that
+the current implementation introduces the minimal number of constraints
+by mashing all possible fields into a super-record. The alternative solution
+(option 2) would introduce additional constraints, when the spec requires us
+to extract an element from a set.
+
+Recall the example with a set of messages:
+
+```tla
+VARIABLES
+  (*
+    @typeAlias: MESSAGE =
+       [ tag: "1a", bal: Int ]
+     | [ tag: "2a", bal: Int, val: Int ]);
+    @type: Set(MESSAGE);
+   *)  
+  msgs
+```  
+
+Consider an existential quantifier over the variable `msgs`:
+
+```tla
+Next ==
+  \E m \in msgs:
+    P
+```
+
+In the current encoding, `m` is a super-record that contains three fields:
+`tag`, `bal`, and `val`, even if some of these fields are not required by the
+actual type of `m`. In the alternative encoding, `m` is a tuple `tup`,
```

Comment:
How is possible for `m` to include all three fields in all cases? E.g., suppose 

```
msgs = {[tag: "1a", bal: 3]}
```

Where do you get your value to associate with the `val` field?

Reply:
If `msgs` has the type `[tag: Str, bal: Int, val]`, which is possible in the current implementation, then the model checker would just introduce a default value for `val`. It's actually not so different from unions on C, where you would have to reserve memory for all fields, even though some of them will never be used.

Reply:
Could you show an example of how you intend to represent a value which has a variant with two different alternatives?

E.g., what would be the equivalent of

```
ev =〈key = ’\t’〉::〈key :: Char | mouse :: Point 〉
```

I think seeing that written out here would help me understand what you envision. I expect adding such an example/comparison might useful for the clarity of the ADR too?

Reply:
You would just introduce a variant over tagged records, instead of having arbitrary types.